loky/backend/resource_tracker.py

```diff
@@ -0,0 +1,277 @@
+                    import re
+                    for i in range(1, len(args)):
+                        args[i] = re.sub("-R+", "-R", args[i])
+                args += ['-c', cmd % (r, VERBOSE)]
+                util.debug("launching resource tracker: {}".format(args))
+                # bpo-33613: Register a signal mask that will block the
+                # signals.  This signal mask will be inherited by the child
+                # that is going to be spawned and will protect the child from a
+                # race condition that can make the child die before it
+                # registers signal handlers for SIGINT and SIGTERM. The mask is
+                # unregistered after spawning the child.
+                try:
+                    if _HAVE_SIGMASK:
+                        signal.pthread_sigmask(signal.SIG_BLOCK,
+                                               _IGNORED_SIGNALS)
+                    pid = spawnv_passfds(exe, args, fds_to_pass)
+                finally:
+                    if _HAVE_SIGMASK:
+                        signal.pthread_sigmask(signal.SIG_UNBLOCK,
+                                               _IGNORED_SIGNALS)
+            except BaseException:
+                os.close(w)
+                raise
+            else:
+                self._fd = w
+                self._pid = pid
+            finally:
+                os.close(r)
+
+    def _check_alive(self):
+        '''Check for the existence of the resource tracker process.'''
+        try:
+            self._send('PROBE', '', 'semlock')
+        except BrokenPipeError:
+            return False
+        else:
+            return True
+
+    def register(self, name, rtype):
+        '''Register a named resource with resource tracker.'''
+        self.ensure_running()
+        self._send('REGISTER', name, rtype)
+
+    def unregister(self, name, rtype):
+        '''Unregister a named resource with resource tracker.'''
+        self.ensure_running()
+        self._send('UNREGISTER', name, rtype)
+
+    def _send(self, cmd, name, rtype):
+        msg = '{0}:{1}:{2}\n'.format(cmd, name, rtype).encode('ascii')
+        if len(name) > 512:
+            # posix guarantees that writes to a pipe of less than PIPE_BUF
+            # bytes are atomic, and that PIPE_BUF >= 512
+            raise ValueError('name too long')
+        nbytes = os.write(self._fd, msg)
+        assert nbytes == len(msg)
+
+
+_resource_tracker = ResourceTracker()
+ensure_running = _resource_tracker.ensure_running
+register = _resource_tracker.register
+unregister = _resource_tracker.unregister
+getfd = _resource_tracker.getfd
+
+
+def main(fd, verbose=0):
+    '''Run semaphore tracker.'''
+    # protect the process from ^C and "killall python" etc
+    signal.signal(signal.SIGINT, signal.SIG_IGN)
+    signal.signal(signal.SIGTERM, signal.SIG_IGN)
+
+    if _HAVE_SIGMASK:
+        signal.pthread_sigmask(signal.SIG_UNBLOCK, _IGNORED_SIGNALS)
```

Comment:
I don't understand why we use both `signal.signal` and the `pthread_sigmask`. This feels redundant to me.. Could you please explain why we need both? Or if we do not, let's stick to `signal.signal` as it works for both `win32` and `unix`.

Reply:
These lines come from a `multiprocessing` patch. It is already introduced in #202. I did not question them, and simply added them for upstream synchronization concerns.

The point of these lines is to avoid race conditions where `SIGINT` or `SIGTERM` signals are sent to the `semaphore_tracker` process before its custom signal handlers are registered in `main`.
`signal.pthread_sigmask` sets a signal mask in the parent process. The [linux docs](http://man7.org/linux/man-pages/man2/sigprocmask.2.html) tells us that a child process created by `fork` inherits the signal mask of its parent. Temporarily setting a mask before the `semaphore_tracker` is launched was the proposed solution to avoid the race condition described above. Note that signal handling modifications via `signal.signal` are not inheritable. So using `signal.signal` instead of `signal.pthread_mask` in these lines is not possible.

Reciprocally, we cannot rely only on `signal.pthread_mask` because this method is not available in `python2` and on `windows`. Would we only support `unix` systems on `python3`, we could simply keep the `_IGNORED_SIGNALS` blocked and suppress those lines AFAICT.